KKUMI-52_token_service

eekrwl opened the pull request:
## 구현내용
### 1. Refresh, Access token 생성 기능
#### * 유효기간
Refresh token은 애플리케이션 서비스인 것을 고려해서 90일로, Access token은 15분으로 설정했습니다.
```java
private final Duration ACCESS_TOKEN_DURATION = Duration.ofMinutes(15);
private final Duration REFRESH_TOKEN_DURATION = Duration.ofDays(90);
```

#### * JWT의 subject는 user의 UUID
토큰의 subject에는 user의 UUID를 넣어주었습니다. 
기존에는 user의 id를 그대로 넣었지만, 아래와 같은 이유로 변경했습니다.

1. 보안성
auto increment되는 Id를 그대로 사용하게 되면 DB에 row가 몇개 있는지 등 유추가 가능해집니다.
보안성을 높이기 위해 UUID를 생성하여 넣어주었습니다.
2. DB 이전 가능성
올해 말쯤 aws 계정을 다른 계정으로 옮겨야 합니다. 옮길 때 auto increment 되는 값은 그대로 유지될지  확언할 수 없다는 조언을 듣고 UUID를 사용하기로 했습니다.

### 2. 포스트 무한스크롤 마지막인지 확인하는 로직 변경
무한스크롤이 끝났는지 확인하기 위해, limit+1 값을 조회하여 limit+1보다 작은 값이 조회되면 끝난 것으로 판단하기로 했습니다.
기존에는 남은 포스트의 개수를 세는 쿼리가 한번 더 나가게 되었는데, 현재 방식으로 변경하면서 쿼리가 한번만 나가게 되었습니다.

eekrwl committed:
KKUMI-52 [FEAT] #19 refresh token entity, repository 생성

refresh token의 entity, repository를 생성했습니다.
보안성과 5개월 뒤 예정된 DB이전을 고려하여 auto increment되는 id로는 충분하지 않을 수있다는 이야기를 듣고 uuid 컬럼을 추가했습니다.

eekrwl committed:
KKUMI-52 [FEAT] #19 토큰 서비스 생성

token service를 생성했습니다.
refresh token과 access token을 생성합니다.
token의 subject로는 user의 uuid를 사용했습니다.
보안성을 높이고, 올해말 다른 DB로 옮겨야하는 일을 고려했습니다.

eekrwl committed:
KKUMI-52 [UPDATE] #19 무한스크롤 마지막 조회 방식 변경

기존: 남은 포스트 몇개인지 개수 셈
변경후: 조회할 포스트+1개 조회하여 포스트+1개 오는지 확인, 이하로 오면 포스트 끝으로 판단

babjo commented on src/main/java/com/swmarastro/mykkumiserver/auth/token/RefreshToken.java:
```diff
@@ -0,0 +1,46 @@
+    public Long id;
+
+    @Column(updatable = false, nullable = false, unique = true, insertable = false)
+    private String uuid;
```
UUID 타입쓰면 그에 맞게 들어갈거에요.

eekrwl replied:
네 수정했습니다!

eekrwl replied:
> UUID 타입쓰면 그에 맞게 들어갈거에요.

멘토님 UUID 관련하여 궁금한 점이 있습니다. uuid를 UUID 타입으로 저장했을 때, String으로 저장하는 것에 비해 장점이 있는지 궁금합니다. 데이터 타입이 안 맞는다든지 해서 에러가 더 많이 나는것 같습니다. 제 생각에는 저장되는 내용자체는 큰 차이가 없을 것 같은데, UUID타입으로 설정했을 때 이점이 있는지 궁금합니당

eekrwl replied:
> > UUID 타입쓰면 그에 맞게 들어갈거에요.
> 
> 멘토님 UUID 관련하여 궁금한 점이 있습니다. uuid를 UUID 타입으로 저장했을 때, String으로 저장하는 것에 비해 장점이 있는지 궁금합니다. 데이터 타입이 안 맞는다든지 해서 에러가 더 많이 나는것 같습니다. 제 생각에는 저장되는 내용자체는 큰 차이가 없을 것 같은데, UUID타입으로 설정했을 때 이점이 있는지 궁금합니당

찾아보니 UTF-8로 인코딩된 문자열은 한 문자당 3-4byte를 차지하기 때문에 저장 공간이 많이 소요된다고 봤습니다. uuid는 -, 알파벳, 숫자 정도로 이뤄져서 저장공간이 많이 필요없을 것 같다는 생각이 듭니다. 하지만 이 부분이 꽤 유의미한 차이가 나는지 궁금합니당..!

babjo replied:
넵넵 작은 프로젝트라 용량신경쓰지 않아도 될거같습니다. UUID 로 저장하면 크기가 절반수준이고 index 비교연산이 작아서 좋긴합니다. 하지만 여기서는 빠르게 만들고 디버깅이 좋으니 문자열 저장도 괜찮을거같습니다.

eekrwl replied:
> 넵넵 작은 프로젝트라 용량신경쓰지 않아도 될거같습니다. UUID 로 저장하면 크기가 절반수준이고 index 비교연산이 작아서 좋긴합니다. 하지만 여기서는 빠르게 만들고 디버깅이 좋으니 문자열 저장도 괜찮을거같습니다.

이것저것 해보다가 DB에는 UUID 타입으로 저장해서 저장공간을 아끼고 서비스 내 자바 코드에서는 string을 사용하는 방식을 사용하니 돌아갑니다 ㅎㅎ 이대로 한번 해보겠습니당!

babjo commented on src/main/java/com/swmarastro/mykkumiserver/user/UserRepository.java:
```diff
@@ -2,5 +2,9 @@
+
 public interface UserRepository extends JpaRepository<User, Long> {
+
+    Optional<User> findByUuid(String uuid);
```
UUID 로 질의하니 관련 index 도 함께 걸어줘요!

eekrwl replied:
넵 인덱스 걸어줬습니다!

babjo commented on src/main/java/com/swmarastro/mykkumiserver/auth/token/TokenService.java:
```diff
@@ -0,0 +1,57 @@
+            throw new CommonException(ErrorCode.INVALID_TOKEN, "유효하지 않은 토큰입니다.", "유효하지 않은 토큰입니다.");
+        }
+        //DB에 존재하는지 확인
+        refreshTokenRepository.findByRefreshToken(refreshToken)
```
refreshToken 전체가 매칭되는 걸 찾으려면 시간이 오래걸릴거같아요!

babjo replied:
저번에 이야기한대로 ID 나 다른 조건으로 빠르게 매칭하는게 좋을거같아ㅛ.

eekrwl replied:
넵 uuid로 찾도록 변경했습니다! 그리고 accessToken 재발급시 refreshToken도 재발급되도록 변경했습니당

babjo replied:
> accessToken 재발급시 refreshToken도 재발급되도록 변경했습니당

refreshToken 으로 accessToken 을 재발행할 때 refreshToken 도 재발행된다일까요?

eekrwl replied:
네 맞습니다 accessToken 재발행시 refreshToken도 재발행되도록 만들어봤습니다..!

babjo replied:
그럼 refresh 를 무한정하는것 아닐까요? refreshtoken 은 다시 재발행하는건 어떤 이유로 일까요?

eekrwl replied:
보안성이 더 좋다는 이야기를 듣고 바꿔봤는데 이렇게까지 할 필요는 없을까요?? 🥹
만약 이 방식대로 하게 된다면 refresh가 매우매우 많아질것 갍아서 지난 토큰들 한번씩 DB에서 지워줄까했습니다..

babjo replied:
> 보안성이 더 좋다는 이야기를 듣고 바꿔봤는데 이렇게까지 할 필요는 없을까요?? 🥹

어떤 측면에서 보안성이 더 좋아질까요??

eekrwl replied:
리프레시 토큰도 한번 사용되면 매번 바뀌어서 재사용할 수 없어서 보안성이 좋지않을까 생각했습니다! 전의 방식대로 하는게 더 나을까용..??

babjo replied:
지금 구현은 재발행해도 이전 refresh token 도 사용할 수 있는 구현 아닐까요?

eekrwl replied:
앗 그러네요....... 원래대로 한번 돌려보겠습니다🥲

eekrwl committed:
KKUMI-52 [UPDATE] #19 uuid String->UUID로 변경

eekrwl committed:
KKUMI_52 [UPDATE] #19 refreshToken 찾을때 uuid 사용하도록 변경, access 재발급시 refresh도 재발급하도록 변경

기존: refreshToken을 DB에서 찾을때 refreshToken으로 찾음
변경후: refreshToken의 uuid로 DB에서 찾음
refreshToken으로 찾으면 토큰이 너무 길어서 uuid로 찾도록 변경했습니다.

기존: accessToken 발급시 refreshToken은 유지
변경후: refreshToken도 함께 재발급
보안성을 높일 수 있습니다.

eekrwl committed:
KKUMI-52 [FIX] #19 UUID 형식 String -> UUID
